olive/passes/onnx/onnx_dag.py

```diff
@@ -0,0 +1,492 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import logging
+from collections import defaultdict
+from enum import Enum
+from pathlib import Path
+from typing import TYPE_CHECKING, Dict, List, Set, Union
+
+import onnx
+from onnx import AttributeProto, GraphProto, NodeProto, TensorProto, ValueInfoProto
+
+from olive.common.config_utils import ConfigBase
+from olive.common.pydantic_v1 import Field
+
+if TYPE_CHECKING:
+    from onnx import ModelProto
+
+logger = logging.getLogger(__name__)
+
+
+class SpecialInput(str, Enum):
+    """Special inputs for ONNX nodes."""
+
+    INPUT = "__input__"  # user input
+    INITIALIZER = "__initializer__"  # constant initializer
+
+
+class SpecialOutput(str, Enum):
+    """Special outputs for ONNX nodes."""
+
+    OUTPUT = "__output__"  # model output
+
+
+class OnnxNode(ConfigBase):
+    """ONNX node."""
+
+    op_type: str
+    inputs: List[str]
+    outputs: List[str]
+    graph_idx: int
+    proto: NodeProto  # reference to the node in the model graph
+
+
+class OnnxIO(ConfigBase):
+    """ONNX input/output.
+
+    Behaves similar to labeled edges in a graph but can connect to multiple nodes.
+    """
+
+    source: str = None
+    destination: List[str] = Field(default_factory=list)
+    graph_idx: int
+    proto: Union[ValueInfoProto, TensorProto] = None
```

Comment:
>ValueInfoProto [](http://example.com/codeflow?start=17&length=14)

could the ProtoBuf type as the pydantic type hint?

Reply:
Yes, it works.

Reply:
Please add comments that it doesn't support serialization.

Reply:
Done.